src/System.Binary.Base64/System/Binary/BufferExtensions.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.Collections.Sequences;
+
+// This should most likely be in System.Text.Primitives and not here. Will move once Base64 moves.
+namespace System.Binary.Base64
+{
+    public static class BufferExtensions
+    {
+        const int stackLength = 32;
+
+        public static void Pipe(this ITransformation transformation, ReadOnlyBytes source, IOutput destination)
+        {
+            int afterMergeSlice = 0;
+            ReadOnlySpan<byte> remainder;
+            Span<byte> stackSpan;
+
+            unsafe
+            {
+                byte* stackBytes = stackalloc byte[stackLength];
```

Comment:
> Unless, if we add a check like if stackSpan.IsEmpty, then we do the allocation.

Can that be addressed outside this PR? I will investigate if it is worth moving it inside the `if (!remainder.IsEmpty)` block within the loop.